skimage/restoration/_nl_means_denoising.pyx

```diff
@@ -0,0 +1,596 @@
+
+    h: float, optional
+        cut-off distance (in gray levels). The higher h, the more permissive
+        one is in accepting patches.
+    """
+    if s % 2 == 0:
+        s += 1  # odd value for symmetric patch
+    cdef int offset = s / 2
+    # Image padding: we need to account for patch size, possible shift,
+    # + 1 for the boundary effects in finite differences
+    cdef int pad_size = offset + d + 1
+    cdef DTYPE_t [:, ::1] padded = np.ascontiguousarray(util.pad(image,
+                                pad_size, mode='reflect').astype(np.float32))
+    cdef DTYPE_t [:, ::1] result = np.zeros_like(padded)
+    cdef DTYPE_t [:, ::1] weights = np.zeros_like(padded)
+    cdef DTYPE_t [:, ::1] integral = np.zeros_like(padded)
+    cdef int n_x, n_y, t1, t2, x, y
+    cdef float weight, distance
+    cdef float alpha
+    cdef float h2 = h ** 2.
+    cdef float s2 = s ** 2.
+    n_x, n_y = image.shape
+    n_x += 2 * pad_size
+    n_y += 2 * pad_size
+    # Outer loops on patch shifts
+    # With t2 >= 0, reference patch is always on the left of test patch
+    for t1 in range(-d, d + 1):
+        for t2 in range(0, d + 1):
+            # alpha is to account for patches on the same column
+            # distance is computed twice in this case
+            if t2 == 0 and t1 is not 0:
+                alpha = 0.5
+            else:
+                alpha = 1.
+            integral = np.zeros_like(padded)
+            for x in range(max(1, - t1), min(n_x, n_x - t1)):
+                for y in range(max(1, - t2), min(n_y, n_y - t2)):
+                    integral[x, y] = (padded[x, y] -
+                                      padded[x + t1, y + t2])**2 + \
+                                      integral[x - 1, y] + integral[x, y - 1] \
+                                      - integral[x - 1, y - 1]
+            for x in range(max(offset, offset - t1),
+                           min(n_x - offset, n_x - offset - t1)):
+                for y in range(max(offset, offset - t2),
+                               min(n_y - offset, n_y - offset - t2)):
+                    distance = integral[x + offset, y + offset] + \
+                               integral[x - offset, y - offset] - \
+                               integral[x - offset, y + offset] - \
+                               integral[x + offset, y - offset]
+                    distance /= (s2 * h2)
+                    # exp of large negative numbers is close to zero
+                    if distance > 5:
+                        continue
+                    weight = alpha * exp(- distance)
+                    weights[x, y] += weight
+                    weights[x + t1, y + t2] += weight
+                    result[x, y] += weight * padded[x + t1, y + t2]
+                    result[x + t1, y + t2] += weight * padded[x, y]
+    for x in range(offset, n_x - offset):
+        for y in range(offset, n_y - offset):
+            # I think there is no risk of division by zero
```

Comment:
Well, maybe I should remove "I think" then. Since there is always a weight of one for the (0, 0) shift, there is no risk of division by zero :-).